### ISSUE TITLE
in ORM REST Data Panache add option to query on fields other than ID ex: GET /people?name=jack

### ISSUE THREAD

**Issue by eliaharm** [quarkusio/quarkus#13520]
**Description**
when using quarkus-hibernate-orm-rest-data-panache it would be good to have the option to query on fields other than ID ex: GET /people?name=jack
I needed it for example to get a list of all cities for a given country but will have to create an additional REST API while using the Panache auto-generated JAX-RS resources.

**Implementation ideas**
none

**Comment by gytis** [comment]
This would be an interesting feature to add. But we should be cautious on what extra features do we add. I don't think it would be a good idea to overcomplicate the extension and try to generate every possible CRUD API variation. 
Adding filtering by column value would make sense IMO. But should we go further and support more complex queries?

**Comment by eliaharm** [comment]
Is it possible in case of more advanced queries to allow overriding a specific method in order to customize it?

**Comment by gytis** [comment]
We don't support overriding yet. An alternative right now would be to not generate a specific method (annotate it with `@ResourceProperties(exposed=false)`) and implement it manually in a different class.

**Comment by fiorenzino** [comment]
Hi Gytis,

something like:

    @ResourceProperties(exposed=false))
    public PanacheQuery<T> getSearch( @Context UriInfo ui) throws Exception
{
         PanacheQuery<T> search;
          ...
     return search;
    }

Should that be possible?

Fiorenzo

Il giorno gio 25 feb 2021 alle ore 12:41 Gytis Trikleris <
notifications@github.com> ha scritto:

**Comment by gytis** [comment]
The exposed is only relevant for the methods that are known to the generator (list, get, add, update, delete). What I meant by using `exposed=false` is to put it on the `list` method (or whichever else you want to override) and then in a separate class implement it manually with the same path.

However, I think in the future we should add a support for the custom implementation that could be added as an interface default method.

**Comment by fiorenzino** [comment]
Hi Gytis,

This is not the first time that someone asked for some customization on
list method.

In your visibility of quarkus project, that feature is in interest for a
limited number of users?

In my company, we developed custom api for that reason and a maven plugin
to simplify the writing of list search(https://github.com/n-essio/queryable)
Could be interesting to contribute to this aspect, to remove our
customization and use only panache api.

Our difficulty is always the writing of Gizmo code, but if someone can
drive us, the effort to contribute will be well spent.

Best Regard

Fiorenzo


Il giorno mar 9 mar 2021 alle ore 09:25 Gytis Trikleris <
notifications@github.com> ha scritto:

**Comment by FroMage** [comment]
Hah, that's interesting. It seems like a useful extension to write. I'm not sure it should come in this particular extension, or be a new extension in Quarkiverse.

I wonder if something like this (https://github.com/n-essio/queryable) already exists, though, perhaps @Sanne or @gavinking or @gsmet or @geoand  know?

TL;DR; it generates `@Filter*` annotations for entity fields annotated with a marker interface, and generates a REST endpoint that supports queries using those fields by applying the required filters.

It reminds me of HS a bit, except this isn't about full-text.

**Comment by Sanne** [comment]
On Tue, 9 Mar 2021 at 14:00, Stéphane Épardaud <notifications@github.com>
wrote:

> Hah, that's interesting. It seems like a useful extension to write. I'm
> not sure it should come in this particular extension, or be a new extension
> in Quarkiverse.
>
> I wonder if something like this (https://github.com/n-essio/queryable)
> already exists, though, perhaps @Sanne <https://github.com/Sanne> or
> @gavinking <https://github.com/gavinking> or @gsmet
> <https://github.com/gsmet> or @geoand <https://github.com/geoand> know?
>
Right, that looks nice. And clearly it exists, you just pointed to it?

If you mean a similar extension, I'm not aware of any, and yes would be
great to see queryable evolve as an extension so that users wouldn't need
to use the Maven plugin, and could benefit from e.g. hot-reload
integrations.

TL;DR; it generates @Filter* annotations for entity fields annotated with a
> marker interface, and generates a REST endpoint that supports queries using
> those fields by applying the required filters.
>
I'm sure I told you before, but yes I do like filters :) Makes it very easy
to manage a complex stack of restrictions which need to be enabled/disabled
according to business rules, and since they can take parameters you can
express many use cases easily.


> It reminds me of HS a bit, except this isn't about full-text.
>
The best part thing is that HS filters and ORM filters can be combined.

**Comment by gavinking** [comment]
@FroMage you might want to support `@NaturalId`. The documentation of this feature is a bit confusing but it might be what the original poster was looking for.

**Comment by Sgitario** [comment]
This was implemented as part of https://github.com/quarkusio/quarkus/pull/29323. Closing.
